pkg/controller/util.go

```diff
@@ -138,55 +137,47 @@ func (c *Controller) getSecretName(pvc *v1.PersistentVolumeClaim) (string, error
 	return name, nil
 }
 
-func (c *Controller) patchPVC(oldData, newData []byte, pvc *v1.PersistentVolumeClaim) error {
-	// patch the pvc clone
-	patch, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1.PersistentVolumeClaim{})
-	if err != nil {
-		return errors.Wrap(err, "error creating pvc patch")
-	}
-	_, err = c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Patch(pvc.Name, types.StrategicMergePatchType, patch)
-	if err != nil {
-		return errors.Wrapf(err, "error patching pvc %q", pvc.Name)
-	}
-	return nil
-}
+// Update and return a copy of the passed-in pvc. Note: the only pvc changes supported are metadata, eg.
+// annotations or labels.
+func (c *Controller) updatePVC(pvc *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	nsName := fmt.Sprintf("%s/%s", pvc.Namespace, pvc.Name)
+	// don't mutate the passed-in pvc since it's likely from the shared informer
+	pvcCopy := pvc.DeepCopy()
+	// save metadata in case Update fails due to stale object
+	metadata := pvcCopy.ObjectMeta
+	var updtPvc *v1.PersistentVolumeClaim
+	var err error
 
-func (c *Controller) clonePVC(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, []byte, error) {
-	pvcClone := claim.DeepCopy()
-	data, err := json.Marshal(pvcClone)
-	if err != nil {
-		return pvcClone, nil, errors.Wrap(err, "error marshalling pvc object")
-	}
-	return pvcClone, data, nil
-}
-
-// Sets an annotation `key: val` in the given PVC.
-// Note: Patch() is used instead of Update() to handle version related field changes.
-func (c *Controller) setPVCAnnotation(pvc *v1.PersistentVolumeClaim, key, val string) error {
-	glog.V(Vdebug).Infof("Adding annotation \"%s: %s\" to pvc \"%s/%s\"\n", key, val, pvc.Namespace, pvc.Name)
+	// loop a few times in case the pvc is stale
+	err = wait.PollImmediate(time.Second*1, time.Second*4, func() (bool, error) {
+		updtPvc, err = c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Update(pvcCopy)
+		if err == nil {
+			return true, nil // successful update
+		}
+		if apierrs.IsConflict(err) { // pvc is likely stale
```

Comment:
better to not retry at all, not even have a loop here. Let the re-queue logic of the pvc work queue handle the retry.